### PR TITLE
Adding Boilerplate Middleware Sections

### PR DESCRIPTION
**Pull request by darthtrevino**
This PR adds 'pre' and 'post' middleware sections. These contain configuration for cross-cutting middleware that is applied to routes matching a precondition.

[pre/post].all - These middleware are applied before/after all defined routes.
[pre/post].[safe/unsafe] - These middleware are applied before/after all safe or unsafe routes. (Safe = GET, OPTIONS, HEAD)
[pre/most].methods.[method] - These methods are applied before/after specific HTTP methods (get, post, etc..)

If a middleware is eligible for multiple sections, they are applied in this order:

pre-all
pre-safe/unsafe
pre-method
<<ROUTE MIDDLEWARE>>
post-method
post-safe/unsafe
post-all

```
{
    pre: {
        all: [],
        safe: [],
        unsafe [],
        methods: {
            get: [],
            // Other HTTP Methods
        }
    },
    // Post is similar
}
```

In addition to these cross-cutting boilerplate config sections, the jefferson library has been refactored into a domain model of small components.

### CODE REVIEW

**Comment by natoverse** [comment]
I like the change from mutable/unmutable to unsafe/safe. Seems more clear. Can you give a little more background on the intention of the PR - the issue doesn't have much description, so I'm not sure what to look for.

File: src/domain/appsection/routes/middleware_composer.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Composes a middleware chain by taking a core chain and adding boilerplate pre and post sections
+ * to it
+ */
+module.exports = {
+    compose (route, conf) {
+        let result = [];
+        let method = route.method.toLowerCase();
+        let isSafe = () => safeMethods[method];
+        let add = (x=[]) => result = result.concat(x);
```
Comment (natoverse) [inline src/domain/appsection/routes/middleware_composer.js]: es6 ftw

File: src/jefferson.boilerplate-middleware.test.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                    ]
+                }
+            }
+        };
+        let app = express();
+        jefferson(app, conf);
+        return checkRequest(request(app).get("/test"), "0123");
+    });
+
+    it("can describe middleware that's invoked before mutable and immutable chains", () => {
```
Comment (natoverse) [inline src/jefferson.boilerplate-middleware.test.js]: unsafe/safe

File: src/jefferson.boilerplate-middleware.test.js [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                unsafe: [middleware.append("M")],
+                method: {
+                    get: [middleware.append(1)],
+                    put: [middleware.append(2)],
+                    post: [middleware.append(3)],
+                    delete: [middleware.append(4)],
+                    options: [middleware.append(5)]
+                }
+            },
+            routes: {
```
Comment (natoverse) [inline src/jefferson.boilerplate-middleware.test.js]: We don't really have a lot of use cases for it, but would it be worth putting 'patch' in here? Also, would it be worthwhile to consistently map verbs to **CRUD** (not including options/head)?

Comment (darthtrevino) [inline src/jefferson.boilerplate-middleware.test.js]: It supports all HTTP methods that express supports. (http://expressjs.com/api.html#app.METHOD). I didn't really want to cover them all here.

Comment (natoverse) [inline src/jefferson.boilerplate-middleware.test.js]: Ok, nm, you just weren't covering all those other methods in the tests - that's fine. I would cover anything that seems likely to be used, but again, I'm not sure *patch* is even going to be common anyway.

**Comment by natoverse** [comment]
'microlibrary', haha. :shipit:

**Pull request merged**